File: 06_strings/06_11_longest_palindromic_subsequence/longest_palindromic_subsequence.js
const extendedPalindrome = (s = '', l = 0, r = 0) => {
    while (l < s.length && r >= 0 && s[l] === s[r]) {
        l++;
        r--;
    }
    return s.substring(r + 1, l);
}

const longestPalindrome = (s = '') => {
    let result = '';
    for (let i = 0; i < s.length; i++) {
        let temp = extendedPalindrome(s, i, i);
        if (temp.length > result.length) {
            result = temp;
        }
        temp = extendedPalindrome(s, i + 1, i);
        if (temp.length > result.length) {
            result = temp;
        }
    }
    return result;
}

// Test cases
console.log('Input: "babad" -> Output: "' + longestPalindrome("babad") + '"');
console.log('Input: "cbbd"  -> Output: "' + longestPalindrome("cbbd") + '"');
console.log('Input: "a"     -> Output: "' + longestPalindrome("a") + '"');
console.log('Input: "ac"    -> Output: "' + longestPalindrome("ac") + '"');

File: 06_strings/06_11_longest_palindromic_subsequence/longest_palindromic_subsequence.py
def extendedPalindrome(s: str, l: int, r: int):
    while l < len(s) and r >= 0 and s[l] == s[r]:
        l += 1
        r -= 1
    return s[r + 1: l]

def longestPalindrome(s: str):
    result = ""
    for i in range(len(s)):
        temp = extendedPalindrome(s, i, i)
        if len(temp) > len(result):
            result = temp
        temp = extendedPalindrome(s, i + 1, i)
        if len(temp) > len(result):
            result = temp
    return result

# Test cases
print("Input: \"babad\" -> Output: \"" + longestPalindrome("babad") + "\"")
print("Input: \"cbbd\"  -> Output: \"" + longestPalindrome("cbbd") + "\"")
print("Input: \"a\"     -> Output: \"" + longestPalindrome("a") + "\"")
print("Input: \"ac\"    -> Output: \"" + longestPalindrome("ac") + "\"")